src/raft_baselines/classifiers/in_context_classifier.py

```diff
@@ -0,0 +1,257 @@
+
+    def render_examples(
+        self,
+        example_dataset: datasets.Dataset,
+        max_tokens_per_example: Optional[int] = None,
+    ) -> str:
+        formatted_examples = [
+            self.format_example(
+                {col: row[col] for col in self.input_cols if col in row},
+                self.class_label_to_string(row[self.class_col]),
+                max_tokens=max_tokens_per_example,
+            )
+            for row in example_dataset
+        ]
+        return self.separator.join(formatted_examples)
+
+    @abstractmethod
+    def semantically_select_training_examples(
+        self, target: Mapping[str, str]
+    ) -> datasets.Dataset:
+        ...
+
+    def select_training_examples(
+        self, target: Mapping[str, str], random_seed: Optional[int] = None
+    ) -> datasets.Dataset:
+        if not self.do_semantic_selection:
+            random.seed(random_seed)
+
+            n_ex = self.num_prompt_training_examples
+            if n_ex is None or len(self.training_data) <= n_ex:
+                return self.training_data
+
+            uniques = defaultdict(lambda: [])
+            for i, row in enumerate(self.training_data):
+                uniques[row["Label"]].append(i)
+
+            indices = []
+            for key in uniques:
+                indices.append(random.choice(uniques[key]))
+            random.shuffle(indices)
+
+            remaining_indices = [
+                i for i in range(len(self.training_data)) if i not in indices
+            ]
+            indices += random.sample(
+                remaining_indices, min(n_ex, len(remaining_indices))
+            )
+
+            return self.training_data.select(indices[:n_ex])
+        else:
+            return self.semantically_select_training_examples(target)
+
+    def format_prompt(
+        self,
+        target: Mapping[str, str],
+        example_dataset: Optional[datasets.Dataset] = None,
+    ) -> str:
+        if self.truncation_params is None:
+            raise ValueError("No truncation strategy provided.")
+
+        max_end_example_tokens, max_train_example_tokens = self.max_example_lengths(
+            len(example_dataset), target
+        )
+        example_str = self.render_examples(
+            example_dataset, max_tokens_per_example=max_train_example_tokens
+        )
+        example_str_and_sep = "" if example_str == "" else example_str + self.separator
+
+        prompt = f"""{self.instructions + self.separator if self.instructions != "" else ""}{example_str_and_sep}{self.format_prompt_end(target, max_tokens=max_end_example_tokens)}"""  # noqa: E501
+        return prompt
+
+    @abstractmethod
+    def _get_raw_probabilities(
+        self,
+        prompt: str,
+    ) -> List[float]:
+        ...
+
+    def _classify_prompt(
+        self,
+        prompt: str,
+    ) -> Dict[str, float]:
+        raw_p = self._get_raw_probabilities(prompt)
+        raw_p = torch.stack(raw_p).cpu().detach().numpy()
```

Comment:
I had to convert the list of tensors to a numpy array for cases when we run the code on a GPU